### PR TITLE
Improved EIP-712 fields validation with optional support

### PR DESCRIPTION
**Pull request by mhofman**
closes: https://linear.app/agoric/issue/AGO-874
closes: https://linear.app/agoric/issue/AGO-1028
refs: https://linear.app/agoric/issue/AGO-681

## Description
Update the orchestration tools handling the generation of, and extraction from, EIP-712 messages to support optional fields, and validate that properties present in the data but not signed over according to the type definition are rejected.

Update the portfolio-api EIP-712 types to leverage the optional fields for `OpenPortfolio` and `SetAutoFeatures` messages.

Ensure that the data is fully validated against the types at runtime, removing viem's `validateTypedData` which does not fully validate all types (and instead relies on the hash logic to detect these issues).

### Security Considerations
Stronger validation of EIP-712 messages

### Scaling Considerations
None. This does the normalization of the types and validation in a single pass. viem's hash logic still does its own validation + hashing in 2 passes.

### Documentation Considerations
The PR introduces a new `TypedDataToStructType` helper used instead of abitype's `TypedDataToPrimitiveTypes`, and which supports optional fields defined at any level. It has some simplifications for edge cases that aren't relevant to our types (circular types).

### Testing Considerations
Added unit tests covering the extra validation.

### Upgrade Considerations
The package updates should be picked up by EMS to enforce the same constraints as on-chain validation. However since the payload itself was pass-through in the first place if it passed validation, there are no changes or coordination required. Updating EMS would let EMS/YDS remove some ad-hoc extra type validation.

### CODE REVIEW

**Comment by mhofman** [comment]
Best reviewed commit-by-commit. The last one is the commit introducing the built-in validation, and could be ejected in a separate PR to reduce the scope.

File: packages/portfolio-api/src/evm-wallet/message-handler-helpers.ts [hunk 1]
```diff
@@ -99,7 +97,6 @@ export const makeEVMHandlerUtils = (viemUtils: {
     isHex,
     hashStruct,
     recoverTypedDataAddress,
-    validateTypedData,
     encodeType,
     getTypesForEIP712Domain,
   ]) {
```
[the remark this answers is not in the document]
Comment (mhofman) [reply packages/portfolio-api/src/evm-wallet/message-handler-helpers.ts]: Pre-existing but fixed.

File: packages/orchestration/src/utils/viem-utils/eip712-normalize.ts [hunk 2]
```diff
@@ -0,0 +1,425 @@
...
+
+/**
+ * A field declared without `optional` starts `required`; one declared
+ * `optional` starts `optional-unseen` and transitions from there as
+ * instances of its struct are visited -- see `visit`.
+ */
+type FieldState = 'optional-unseen' | 'optional-seen' | 'required';
+
+/** A struct's fields, keyed by name, tracking each field's {@link FieldState}. */
+type TypeFields = Record<string | number, { state: FieldState; type: string }>;
```
Comment (gibson042) [inline packages/orchestration/src/utils/viem-utils/eip712-normalize.ts]: Under what circumstances will `number` come up? I'm wondering if the key can be simplified to just `string`.

File: packages/orchestration/src/utils/viem-utils/eip712-normalize.ts [hunk 3]
```diff
@@ -0,0 +1,425 @@
...
+    // primitive with a matching JS value; also catches an unknown type.
+    assertValidPrimitive(fieldType, value);
+    return value;
+  }
+
+  // A declared struct or array needs an actual object/array value.
+  if (typeof value !== 'object' || value === null) {
+    throw new Error(
+      `Expected ${isArray ? 'an array' : 'an object'} for EIP-712 type "${truncate(fieldType)}", got ${describeValue(value)}`,
+    );
```
Comment (gibson042) [inline packages/orchestration/src/utils/viem-utils/eip712-normalize.ts]: Suggestion for error message greppability:
```suggestion
    if (isArray) {
      throw new Error(
        `Expected an array for EIP-712 type "${truncate(fieldType)}", got ${describeValue(value)}`,
      );
    } else {
      throw new Error(
        `Expected an object for EIP-712 type "${truncate(fieldType)}", got ${describeValue(value)}`,
      );
    }
```

File: packages/orchestration/src/utils/viem-utils/eip712-normalize.ts [hunk 4]
```diff
@@ -0,0 +1,425 @@
...
+    : str;
+
+/**
+ * A short, bounded description of a value for error messages. Only
+ * `string`/`number`/`bigint` can be arbitrarily long themselves; anything
+ * else is described by shape/type rather than rendered.
+ */
+const describeValue = (value: unknown): string => {
+  if (value === null) return 'null';
+  if (typeof value === 'string') return truncate(JSON.stringify(value));
```
Comment (gibson042) [inline packages/orchestration/src/utils/viem-utils/eip712-normalize.ts]: We should avoid unbalanced output like `"Lorem ipsum...($n chars total)`.
```suggestion
  if (typeof value === 'string') {
    if (str.length > MAX_DESCRIBED_LENGTH) {
      const truncated = `${value.slice(0, MAX_DESCRIBED_LENGTH)}...`;
      return `${JSON.stringify(truncated)}(${value.length} chars total)`;
    }
    return JSON.stringify(value);
  }
```

File: packages/orchestration/src/utils/viem-utils/eip712-normalize.ts [hunk 5]
```diff
@@ -0,0 +1,425 @@
...
+  if (typeof value === 'bigint') return truncate(`${value}n`);
```
Comment (gibson042) [inline packages/orchestration/src/utils/viem-utils/eip712-normalize.ts]:
```suggestion
  if (typeof value === 'bigint') return `${truncate(`${value}`)}n`;
```

File: packages/orchestration/src/utils/viem-utils/eip712-normalize.ts [hunk 6]
```diff
@@ -0,0 +1,425 @@
...
+      throw new Error(
+        `Unexpected field(s) on EIP-712 type "${truncate(fieldType)}": ${truncate(extraKeys.join(', '))}`,
+      );
+    }
+  }
+
+  // `fieldName in obj`, not `hasOwnProperty`: matches how viem itself reads
+  // field values (plain property access resolves the prototype chain).
+  // `type` is null-prototype or a fresh array-literal record, so there's no
+  // inherited-property risk to guard against here.
```
Comment (gibson042) [inline packages/orchestration/src/utils/viem-utils/eip712-normalize.ts]:
```suggestion
  // `type` must be null-prototype or a fresh array-literal record (so there's
  // no inherited-property risk to guard against here), but this check protects
  // against future refactorings.
  if (Object.getPrototypeOf(type) && !Array.isArray(type)) {
    throw new Error(
      `EIP-712 type "${truncate(fieldType)}" must be described by a null-prototype object or an array`,
    );
  }
```

File: packages/orchestration/src/utils/viem-utils/eip712-normalize.ts [hunk 7]
```diff
@@ -0,0 +1,425 @@
...
+    // Not a declared struct or array, so it must be a recognized Solidity
+    // primitive with a matching JS value; also catches an unknown type.
+    assertValidPrimitive(fieldType, value);
+    return value;
+  }
+
+  // A declared struct or array needs an actual object/array value.
+  if (typeof value !== 'object' || value === null) {
+    throw new Error(
+      `Expected ${isArray ? 'an array' : 'an object'} for EIP-712 type "${truncate(fieldType)}", got ${describeValue(value)}`,
```
Comment (gibson042) [inline packages/orchestration/src/utils/viem-utils/eip712-normalize.ts]: These message strings look wrong to me; `truncate(string)` can return arbitrary contents that break out of these enclosing quotes. Maybe it needs JSON.stringify as well?

**Review by gibson042 — APPROVED** [review]
packages/portfolio-api/src/evm-wallet/message-handler-helpers.ts is a bit unweildy and still seems like it could be DRYed out, but in general I like this modulo some minor suggestions.

File: packages/orchestration/src/utils/viem-utils/eip712-normalize.ts (hunk 2 above)
Comment (mhofman) [reply packages/orchestration/src/utils/viem-utils/eip712-normalize.ts]: Array types end up creating a synthetic type where the keys are the indices.

File: packages/orchestration/src/utils/permit2/signatureTransferHelpers.ts [hunk 8]
```diff
@@ -170,7 +170,12 @@ export function validatePermit2Domain(
     }
   }
 
-  // XXX: check no extra fields?
+  const extraKeys = Object.keys(extra);
+  if (extraKeys.length) {
+    throw new Error(
+      `Unexpected field(s) in permit2 domain: ${extraKeys.join(', ')}`,
+    );
+  }
```
[the remark this answers is not in the document]
Comment (mhofman) [reply packages/orchestration/src/utils/permit2/signatureTransferHelpers.ts:178]: The strings of all these keys already have to be allocated, so allocating a join of them doesn't create a significant risk.
